src/Orchard.Environment.Extensions/ExtensionManager.cs

```diff
@@ -307,5 +321,22 @@ private static string GetSourceFeatureNameForType(Type type, string extensionId)
             }
             return extensionId;
         }
+
+        private class LazyConcurrentDictionary<TKey, TValue>
+        {
+            private readonly ConcurrentDictionary<TKey, Lazy<TValue>> concurrentDictionary;
+
+            public LazyConcurrentDictionary()
+            {
+                this.concurrentDictionary = new ConcurrentDictionary<TKey, Lazy<TValue>>();
+            }
+
+            public TValue GetOrAdd(TKey key, Func<TKey, TValue> valueFactory)
+            {
+                var lazyResult = this.concurrentDictionary.GetOrAdd(key, k => new Lazy<TValue>(() => valueFactory(k), LazyThreadSafetyMode.ExecutionAndPublication));
```

Comment:
Added comment, its to do with the valuefactory not being thread safe https://blogs.endjin.com/2015/10/using-lazy-and-concurrentdictionary-to-ensure-a-thread-safe-run-once-lazy-loaded-collection/